src/ripple/app/paths/impl/DirectStep.cpp

```diff
@@ -835,18 +835,26 @@ DirectStepI<TDerived>::qualityUpperBound(ReadView const& v, DebtDirection& dir)
 {
     auto const prevStepDebtDir = dir;
     dir = this->debtDirection(v, StrandDirection::forward);
-    std::uint32_t const srcQOut = [&]() -> std::uint32_t {
-        if (redeems(prevStepDebtDir) && issues(dir))
-            return transferRate(v, src_).value;
-        return QUALITY_ONE;
+
+    auto const [srcQOut, dstQIn] = [&] {
+        if (dir == DebtDirection::redeems)
+        {
+            return qualitiesSrcRedeems(v);
+        }
+        else
+        {
+            return qualitiesSrcIssues(v, prevStepDebtDir);
+        }
     }();
-    auto dstQIn =
-        static_cast<TDerived const*>(this)->quality(v, QualityDirection::in);
 
-    if (isLast_ && dstQIn > QUALITY_ONE)
-        dstQIn = QUALITY_ONE;
     Issue const iss{currency_, src_};
-    return Quality(getRate(STAmount(iss, srcQOut), STAmount(iss, dstQIn)));
+    // Be careful not to switch the parameters to `getRate`. The
+    // `getRate(offerOut, offerIn)` function is usually used for offers. It
+    // returns offerIn/offerOut. For a direct step, the rate is dstQOut/srcQIn
```

Comment:
Small but important mixup:

> For a direct step, the rate is ~dstQOut/srcQIn~ srcQOut/dstQIn

Reply:
Nice catch! Fixed in 593ef30bf `[fold] Fix type and use std::minmax`